File: src/analysis.ts
import type { StrokeData, Workout } from "./models.ts";
import { formatSeconds, isIntervalWorkout, pace500mSeconds, parsedDate } from "./models.ts";

const TENTHS_PER_SECOND = 10;
const PACE_DISTANCE = 500;

export interface SplitRow {
  index: number;
  distance: number | null;
  time_seconds: number;
  pace_500m_seconds: number | null;
  pace_500m: string | null;
  stroke_rate: number | null;
  hr_avg: number | null;
  hr_max: number | null;
}

export type SplitShape = "even" | "negative" | "positive" | "variable" | "unknown";

const SHAPE_DRIFT_THRESHOLD_SECONDS = 1.5;
const SHAPE_SPREAD_THRESHOLD_SECONDS = 6;

export function splitTable(w: Workout): SplitRow[] {
  const splits = w.workout?.splits ?? [];
  return splits.map((s, i) => {
    const seconds = s.time / TENTHS_PER_SECOND;
    const pace =
      s.distance != null && s.distance > 0 ? seconds * (PACE_DISTANCE / s.distance) : null;
    return {
      index: i + 1,
      distance: s.distance ?? null,
      time_seconds: Math.round(seconds * 10) / 10,
      pace_500m_seconds: pace != null ? Math.round(pace * 10) / 10 : null,
      pace_500m: pace != null ? formatSeconds(pace) : null,
      stroke_rate: s.stroke_rate ?? null,
      hr_avg: s.heart_rate?.average ?? null,
      hr_max: s.heart_rate?.max ?? null,
    };
  });
}

export function splitShape(rows: SplitRow[]): SplitShape {
  const paces = rows.map((r) => r.pace_500m_seconds).filter((p): p is number => p != null);
  if (paces.length < 2) return "unknown";

  const mid = Math.floor(paces.length / 2);
  const firstHalf = paces.slice(0, mid);
  const secondHalf = paces.slice(paces.length - mid);
  const avg = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
  const drift = avg(firstHalf) - avg(secondHalf);
  const spread = Math.max(...paces) - Math.min(...paces);

  if (Math.abs(drift) <= SHAPE_DRIFT_THRESHOLD_SECONDS) {
    return spread > SHAPE_SPREAD_THRESHOLD_SECONDS ? "variable" : "even";
  }
  return drift > 0 ? "negative" : "positive";
}

export interface StrokeSummary {
  samples: number;
  avg_pace_500m_seconds: number | null;
  avg_pace_500m: string | null;
  avg_spm: number | null;
  avg_hr: number | null;
  max_hr: number | null;
}

export function strokeSummary(strokes: StrokeData[]): StrokeSummary {
  let paceSum = 0;
  let paceCount = 0;
  let spmSum = 0;
  let spmCount = 0;
  let hrSum = 0;
  let hrCount = 0;
  let maxHR = 0;
  for (const s of strokes) {
    if (s.p != null && s.p > 0) {
      paceSum += s.p / TENTHS_PER_SECOND;
      paceCount++;
    }
    if (s.spm != null && s.spm > 0) {
      spmSum += s.spm;
      spmCount++;
    }
    if (s.hr != null && s.hr > 0) {
      hrSum += s.hr;
      hrCount++;
      if (s.hr > maxHR) maxHR = s.hr;
    }
  }
  const avgPace = paceCount > 0 ? paceSum / paceCount : null;
  return {
    samples: strokes.length,
    avg_pace_500m_seconds: avgPace != null ? Math.round(avgPace * 10) / 10 : null,
    avg_pace_500m: avgPace != null ? formatSeconds(avgPace) : null,
    avg_spm: spmCount > 0 ? Math.round((spmSum / spmCount) * 10) / 10 : null,
    avg_hr: hrCount > 0 ? Math.round(hrSum / hrCount) : null,
    max_hr: maxHR > 0 ? maxHR : null,
  };
}

export interface HRPaceBand {
  band_start_seconds: number;
  band: string;
  workouts: number;
  avg_hr: number;
  early_avg_hr: number | null;
  late_avg_hr: number | null;
  hr_delta: number | null;
}

const BAND_WIDTH_SECONDS = 5;
const MIN_STEADY_DISTANCE = 1000;

export function hrAtPace(workouts: Workout[], now: Date, weeks: number): HRPaceBand[] {
  const cutoff = new Date(now);
  cutoff.setDate(cutoff.getDate() - weeks * 7);
  const midpoint = new Date(cutoff.getTime() + (now.getTime() - cutoff.getTime()) / 2);

  const steady = workouts.filter((w) => {
    if (isIntervalWorkout(w)) return false;
    if (w.distance < MIN_STEADY_DISTANCE) return false;
    if (!w.heart_rate?.average || w.heart_rate.average <= 0) return false;
    const t = parsedDate(w);
    return t >= cutoff && t <= now && pace500mSeconds(w) > 0;
  });

  const bands = new Map<number, { all: number[]; early: number[]; late: number[] }>();
  for (const w of steady) {
    const pace = pace500mSeconds(w);
    const bandStart = Math.floor(pace / BAND_WIDTH_SECONDS) * BAND_WIDTH_SECONDS;
    if (!bands.has(bandStart)) bands.set(bandStart, { all: [], early: [], late: [] });
    const bucket = bands.get(bandStart)!;
    const hr = w.heart_rate!.average!;
    bucket.all.push(hr);
    if (parsedDate(w) < midpoint) bucket.early.push(hr);
    else bucket.late.push(hr);
  }

  const avg = (xs: number[]) =>
    xs.length > 0 ? Math.round(xs.reduce((a, b) => a + b, 0) / xs.length) : null;

  return [...bands.entries()]
    .sort(([a], [b]) => a - b)
    .map(([bandStart, bucket]) => {
      const early = avg(bucket.early);
      const late = avg(bucket.late);
      return {
        band_start_seconds: bandStart,
        band: `${formatSeconds(bandStart)}–${formatSeconds(bandStart + BAND_WIDTH_SECONDS)}`,
        workouts: bucket.all.length,
        avg_hr: avg(bucket.all)!,
        early_avg_hr: early,
        late_avg_hr: late,
        hr_delta: early != null && late != null ? late - early : null,
      };
    });
}

File: internal/analysis/analysis.go
package analysis

import (
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/richhaase/c2/internal/models"
)

const (
	paceDistance                = 500
	shapeDriftThresholdSeconds  = 1.5
	shapeSpreadThresholdSeconds = 6
	bandWidthSeconds            = 5
	minSteadyDistance           = 1000
)

type SplitRow struct {
	Index           int      `json:"index"`
	Distance        *int     `json:"distance"`
	TimeSeconds     float64  `json:"time_seconds"`
	Pace500mSeconds *float64 `json:"pace_500m_seconds"`
	Pace500m        *string  `json:"pace_500m"`
	StrokeRate      *int     `json:"stroke_rate"`
	HRAvg           *int     `json:"hr_avg"`
	HRMax           *int     `json:"hr_max"`
}

type Shape string

const (
	ShapeEven     Shape = "even"
	ShapeNegative Shape = "negative"
	ShapePositive Shape = "positive"
	ShapeVariable Shape = "variable"
	ShapeUnknown  Shape = "unknown"
)

func roundTenth(x float64) float64 {
	return math.Round(x*10) / 10
}

func SplitTable(w models.Workout) []SplitRow {
	var splits []models.WorkoutSplit
	if w.Workout != nil {
		splits = w.Workout.Splits
	}
	rows := make([]SplitRow, 0, len(splits))
	for i, s := range splits {
		seconds := s.Time / models.TenthsPerSecond
		row := SplitRow{
			Index:       i + 1,
			Distance:    s.Distance,
			TimeSeconds: roundTenth(seconds),
			StrokeRate:  s.StrokeRate,
		}
		if s.Distance != nil && *s.Distance > 0 {
			pace := seconds * (paceDistance / float64(*s.Distance))
			rounded := roundTenth(pace)
			formatted := models.FormatSeconds(pace)
			row.Pace500mSeconds = &rounded
			row.Pace500m = &formatted
		}
		if s.HeartRate != nil {
			row.HRAvg = s.HeartRate.Average
			row.HRMax = s.HeartRate.Max
		}
		rows = append(rows, row)
	}
	return rows
}

func SplitShape(rows []SplitRow) Shape {
	paces := make([]float64, 0, len(rows))
	for _, r := range rows {
		if r.Pace500mSeconds != nil {
			paces = append(paces, *r.Pace500mSeconds)
		}
	}
	if len(paces) < 2 {
		return ShapeUnknown
	}

	mid := len(paces) / 2
	firstHalf := paces[:mid]
	secondHalf := paces[len(paces)-mid:]
	drift := mean(firstHalf) - mean(secondHalf)

	lowest, highest := paces[0], paces[0]
	for _, p := range paces {
		if p < lowest {
			lowest = p
		}
		if p > highest {
			highest = p
		}
	}
	spread := highest - lowest

	if math.Abs(drift) <= shapeDriftThresholdSeconds {
		if spread > shapeSpreadThresholdSeconds {
			return ShapeVariable
		}
		return ShapeEven
	}
	if drift > 0 {
		return ShapeNegative
	}
	return ShapePositive
}

func mean(xs []float64) float64 {
	sum := 0.0
	for _, x := range xs {
		sum += x
	}
	return sum / float64(len(xs))
}

type StrokeSummaryResult struct {
	Samples            int      `json:"samples"`
	AvgPace500mSeconds *float64 `json:"avg_pace_500m_seconds"`
	AvgPace500m        *string  `json:"avg_pace_500m"`
	AvgSPM             *float64 `json:"avg_spm"`
	AvgHR              *int     `json:"avg_hr"`
	MaxHR              *float64 `json:"max_hr"`
}

func StrokeSummary(strokes []models.StrokeData) StrokeSummaryResult {
	paceSum, spmSum, hrSum := 0.0, 0.0, 0.0
	paceCount, spmCount, hrCount := 0, 0, 0
	maxHR := 0.0
	for _, s := range strokes {
		if s.P != nil && *s.P > 0 {
			paceSum += *s.P / models.TenthsPerSecond
			paceCount++
		}
		if s.SPM != nil && *s.SPM > 0 {
			spmSum += *s.SPM
			spmCount++
		}
		if s.HR != nil && *s.HR > 0 {
			hrSum += *s.HR
			hrCount++
			if *s.HR > maxHR {
				maxHR = *s.HR
			}
		}
	}

	out := StrokeSummaryResult{Samples: len(strokes)}
	if paceCount > 0 {
		avgPace := paceSum / float64(paceCount)
		rounded := roundTenth(avgPace)
		formatted := models.FormatSeconds(avgPace)
		out.AvgPace500mSeconds = &rounded
		out.AvgPace500m = &formatted
	}
	if spmCount > 0 {
		avgSPM := roundTenth(spmSum / float64(spmCount))
		out.AvgSPM = &avgSPM
	}
	if hrCount > 0 {
		avgHR := int(math.Round(hrSum / float64(hrCount)))
		out.AvgHR = &avgHR
	}
	if maxHR > 0 {
		out.MaxHR = &maxHR
	}
	return out
}

type HRPaceBand struct {
	BandStartSeconds int    `json:"band_start_seconds"`
	Band             string `json:"band"`
	Workouts         int    `json:"workouts"`
	AvgHR            int    `json:"avg_hr"`
	EarlyAvgHR       *int   `json:"early_avg_hr"`
	LateAvgHR        *int   `json:"late_avg_hr"`
	HRDelta          *int   `json:"hr_delta"`
}

type hrBucket struct {
	all   []int
	early []int
	late  []int
}

func HRAtPace(workouts []models.Workout, now time.Time, weeks int) []HRPaceBand {
	cutoff := now.AddDate(0, 0, -weeks*7)
	midpoint := cutoff.Add(now.Sub(cutoff) / 2)

	buckets := map[int]*hrBucket{}
	for _, w := range workouts {
		if models.IsIntervalWorkout(w) {
			continue
		}
		if w.Distance < minSteadyDistance {
			continue
		}
		if w.HeartRate == nil || w.HeartRate.Average == nil || *w.HeartRate.Average <= 0 {
			continue
		}
		at := models.ParsedDate(w)
		if at.Before(cutoff) || at.After(now) {
			continue
		}
		pace := models.Pace500mSeconds(w)
		if pace <= 0 {
			continue
		}

		bandStart := int(math.Floor(pace/bandWidthSeconds)) * bandWidthSeconds
		bucket := buckets[bandStart]
		if bucket == nil {
			bucket = &hrBucket{}
			buckets[bandStart] = bucket
		}
		hr := *w.HeartRate.Average
		bucket.all = append(bucket.all, hr)
		if at.Before(midpoint) {
			bucket.early = append(bucket.early, hr)
		} else {
			bucket.late = append(bucket.late, hr)
		}
	}

	starts := make([]int, 0, len(buckets))
	for start := range buckets {
		starts = append(starts, start)
	}
	sort.Ints(starts)

	bands := make([]HRPaceBand, 0, len(starts))
	for _, start := range starts {
		bucket := buckets[start]
		early := meanRounded(bucket.early)
		late := meanRounded(bucket.late)
		band := HRPaceBand{
			BandStartSeconds: start,
			Band: fmt.Sprintf("%s–%s",
				models.FormatSeconds(float64(start)),
				models.FormatSeconds(float64(start+bandWidthSeconds))),
			Workouts:   len(bucket.all),
			AvgHR:      *meanRounded(bucket.all),
			EarlyAvgHR: early,
			LateAvgHR:  late,
		}
		if early != nil && late != nil {
			delta := *late - *early
			band.HRDelta = &delta
		}
		bands = append(bands, band)
	}
	return bands
}

func meanRounded(xs []int) *int {
	if len(xs) == 0 {
		return nil
	}
	sum := 0
	for _, x := range xs {
		sum += x
	}
	v := int(math.Round(float64(sum) / float64(len(xs))))
	return &v
}
